x-pack/plugins/aiops/public/components/log_categorization/log_categorization_for_embeddable/log_categorization_for_discover.tsx

```diff
@@ -0,0 +1,472 @@
+    randomSampler,
+  } = useCategorizeRequest(randomSamplerStorage);
+  const [stateFromUrl] = usePageUrlState<LogCategorizationPageUrlState>(
+    'logCategorization',
+    getDefaultLogCategorizationAppState({
+      searchQuery: createMergedEsQuery(query, filters, dataView, uiSettings),
+    })
+  );
+  const [selectedCategories, setSelectedCategories] = useState<Category[]>([]);
+  const [selectedField, setSelectedField] = useState<DataViewField | null>(null);
+  const [fields, setFields] = useState<DataViewField[]>([]);
+  const [currentDocumentStatsHash, setCurrentDocumentStatsHash] = useState<number | null>(null);
+  const [previousDocumentStatsHash, setPreviousDocumentStatsHash] = useState<number>(0);
+  const [currentAdditionalConfigsHash, setCurrentAdditionalConfigsHash] = useState<number | null>(
+    null
+  );
+  const [previousAdditionalConfigsHash, setPreviousAdditionalConfigsHash] = useState<number | null>(
+    null
+  );
+  const [loading, setLoading] = useState<boolean | null>(null);
+  const [eventRate, setEventRate] = useState<EventRate>([]);
+  const [data, setData] = useState<{
+    categories: Category[];
+    displayExamples: boolean;
+    totalCategories: number;
+  } | null>(null);
+  const [fieldValidationResult, setFieldValidationResult] = useState<FieldValidationResults | null>(
+    null
+  );
+  const tableState = useTableState<Category>([], 'key');
+
+  useEffect(
+    function initFields() {
+      setCurrentDocumentStatsHash(null);
+      setSelectedField(null);
+      setLoading(null);
+      const { dataViewFields, messageField } = getMessageField(dataView);
+      setFields(dataViewFields);
+      setSelectedField(messageField);
+    },
+    [dataView]
+  );
+
+  const cancelRequest = useCallback(() => {
+    cancelWiderTimeRangeRequest();
+    cancelValidationRequest();
+    cancelCategorizationRequest();
+  }, [cancelCategorizationRequest, cancelValidationRequest, cancelWiderTimeRangeRequest]);
+
+  useEffect(
+    function cancelRequestOnLeave() {
+      mounted.current = true;
+      return () => {
+        mounted.current = false;
+        cancelRequest();
+      };
+    },
+    [cancelRequest, mounted]
+  );
+
+  const { searchQuery } = useSearch(
+    { dataView, savedSearch: savedSearch ?? null },
+    stateFromUrl,
+    true
+  );
+
+  const { documentStats, timefilter, earliest, latest, intervalMs, forceRefresh } = useData(
+    dataView,
+    'log_categorization',
+    searchQuery,
+    () => {},
+    undefined,
+    undefined,
+    BAR_TARGET,
+    false
+  );
+
+  const onAddFilter = useCallback(
+    (values: Filter, alias?: string) => {
+      const filter = buildEmptyFilter(false, dataView.id);
+      if (alias) {
+        filter.meta.alias = alias;
+      }
+      filter.query = values.query;
+      if (typeof input.switchToDocumentView === 'function') {
+        input.switchToDocumentView();
+      }
+      setTimeout(() => {
+        filterManager.addFilters([filter]);
+      }, 500);
```

Comment:
It's a work around to solve a recent bug where switching to the documents tab would fail. The filters need to be added slightly after the call to switch tabs.
I appreciate using a delay like this is a bit hacky, but I could not find a better solution, from my testing this does not fail.

Reply:
Discover's appState has a `subscribe` method. You can try subscribing to a state change and adding filters on the `viewMode` update.

Reply:
Good idea. Updated in https://github.com/elastic/kibana/pull/186539/commits/033e8e3b3af507c64c5886c3105cac251218f6e6